In [480]:
import numpy as np
import pandas as pd

#raw data as a csv file
filepath = 'data/NBASTATS2023.csv' 

#display all rows/columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

#read csv into specific dataframe
df = pd.read_csv(filepath)

#shows how many players and fields we are working with

print(df.shape) 

#uncomment below to see full dataset
#print(df)

#check for missing values in our dataframe
#ensure dataset was formatted and parsed ok

missing_values = df.isnull().sum()

print(missing_values)

#retrieve values for a specific player
print("Bam Adebayo Stats \n")
df.loc[df['NAME'] == 'Bam Adebayo']




(609, 29)
RANK     609
NAME       0
TEAM       0
POS        0
AGE        0
GP         0
MPG        0
USG%       0
TO%        1
FTA        0
FT%        0
2PA        0
2P%        0
3PA        0
3P%        0
eFG%       3
TS%        3
PPG        0
RPG        0
APG        0
SPG        0
BPG        0
TPG        0
P+R        0
P+A        0
P+R+A      0
VI         0
ORtg      24
DRtg      23
dtype: int64
Bam Adebayo Stats 



RANK         NAME TEAM  POS   AGE  GP   MPG  USG%   TO%  FTA    FT%   2PA  \
2   NaN  Bam Adebayo  Mia  C-F  25.7  75  34.6  25.3  14.4  402  0.806  1102   

     2P%  3PA    3P%   eFG%    TS%   PPG  RPG  APG  SPG  BPG  TPG   P+R   P+A  \
2  0.545   12  0.083  0.541  0.592  20.4  9.2  3.2  1.2  0.8  2.5  29.6  23.6   

   P+R+A   VI   ORtg   DRtg  
2   32.8  9.8  116.0  105.6

CLEANING

We can see that rank shows as invalid(NaN) for all players(if we uncomment line 19) , so lets drop rank from our dataset

We also can do further cleaning of our dataset. In this case, players whos stats cannot be calculated for some fields(NaN) likely have played not enough games, or have not gotten >0 in a certain stat. We can trim our dataset by either filtering players out who have played very few minutes/games(option 1), or we can just remove any player that has a null value for any stat(option 2), or input filler numbers for our NaN values(option 3). If you dont know exactly how some of the advanced stats are calculated, option 3 is risky, so we will focus on the first two options

In [482]:
df = df.drop('RANK', axis=1)
#show new shape, should have 1 less column than previously
df.shape

(609, 28)

In [483]:


#Option 1
#removes players who ahve played less than 10 games and average less than 
#minutes per game
option_1_df = df.loc[(df['GP']>10) & (df['MPG']>5)]
print(option_1_df.shape)

missing_values = option_1_df.isnull().sum()

print(missing_values)



(503, 28)
NAME     0
TEAM     0
POS      0
AGE      0
GP       0
MPG      0
USG%     0
TO%      0
FTA      0
FT%      0
2PA      0
2P%      0
3PA      0
3P%      0
eFG%     0
TS%      0
PPG      0
RPG      0
APG      0
SPG      0
BPG      0
TPG      0
P+R      0
P+A      0
P+R+A    0
VI       0
ORtg     0
DRtg     0
dtype: int64


We can see that we have trimmed 106 players from our list using the minutes/games restriction, and have in fact gotten
rid of all the players with missing values. However, this is likely overkill as some players have played less than 10 games and less than 5 minutes per game while having analyzable values.


In [485]:
#Option 2
#remove only players who have an invalid value
o2_df= df.loc[(df['ORtg'].notnull()) & (df['eFG%'].notnull())]
print(o2_df.shape)
#we can add in all the other or clauses, but these 2 do the trick due to how these stats overlap

missing_values = o2_df.isnull().sum()

print(missing_values)

(585, 28)
NAME     0
TEAM     0
POS      0
AGE      0
GP       0
MPG      0
USG%     0
TO%      0
FTA      0
FT%      0
2PA      0
2P%      0
3PA      0
3P%      0
eFG%     0
TS%      0
PPG      0
RPG      0
APG      0
SPG      0
BPG      0
TPG      0
P+R      0
P+A      0
P+R+A    0
VI       0
ORtg     0
DRtg     0
dtype: int64


Now we can see that we have elimnated only 24 players and still have all analyzable values

In [487]:
#I would like to see players who score over 20 points per game
#lets store this as a new dataframe and we can do some transformations
pts_over_20 = o2_df.loc[o2_df['PPG'] >= 20]
print(pts_over_20.shape)

#From this we can see that there are 60 players in the 2022-23 NBA season who averaged over 20 points each game



(60, 28)


To do some more advanced filtering: 
I want to separate elite shooters(1) from strong defensive
players(2), from those who are neither(3), or both(4). We are going to create a new field to tag our players with, and set some conditions for each category. 

In order to qualify as an elite shooter:
- over 50 games played(large sample size,82 games is max)
- score over 20 points per game(ppg)
- High efficiency(Free throw% over 80%, Field goal percentage over 35%, True shooting % over 55%)

In order to qualify as a strong defensive player:
- over 50 games played
- 107 or lower defensive rating(league average is 110, lower is better)
- 1 block or steal per game




In [580]:
elite_shooters = pts_over_20.loc[(pts_over_20['GP']>50) & (pts_over_20['TS%']> .55) 
                                 & (pts_over_20['eFG%'] >.35) & (pts_over_20['FT%']>.80)]
print(elite_shooters.shape)
#elite_shooters

(31, 28)


We can see we have a pool of 32 players who meet our criteria for elite shooters

In [582]:
strong_defense = o2_df.loc[o2_df['GP']>20]
strong_defense = strong_defense.loc[(strong_defense['DRtg']<107) 
                                    & ((strong_defense['SPG']>=1)|(strong_defense['BPG']>=1))]
print(strong_defense.shape)
#strong_defense

(42, 29)


We have 47 players who meet our criteria as strong defenders

Next lets move our criteria to value for fantasy games. Fantasy basketball for the league I am in uses the following modifiers: 

Points Scored:1 Points

Total Rebounds:1.2 Points

Assists:1.5 Points

Steals:3 Points

Blocked Shots:3 Points

Turnovers:-1 Points

Missed shot = -.3 Points

(in order to accurately calculate efficiency based on points scored/missed shots, we will use a weighted formula based on accuracy of 2 pointers, 3 pointers, and freethrows)

In [494]:
#Lets create a fantasy points category and append it to our dataframe, maxing out at two decimal points

from math import trunc

#truncate our values so that its rounded to 2 decimals
#num: float to be truncated
def truncate(num):
    num = trunc(num*100)/100
    return num

#df: pandas dataframe containing NBA seasonal data from yahoo
#mult_list: a list containing float multipliers in order of reb, assists, steals, blocks, turnovers, missedshot
def fanpoint_calculator(df, mult_list):
    fp = []
    for index, row in df.iterrows():
        fanpoints = truncate((mult_list[0] * row['RPG']) + (mult_list[1] * row['APG']) +
                             (mult_list[2]*row['SPG']) +(mult_list[3] * row['BPG']) -(mult_list[4] * row['TPG']) + 
                             #formula calculating points scored and missed shots
                             (2*row['2P%']) +(3*row['3P%']) -(mult_list[5]*(1-(row['2P%'])+(1-row['3P%']))) +(row['FTA']*row['FT%']/row['GP']))
    
        fp.append(fanpoints)
    o2_df['FANPOINTS'] = fp


  


In [495]:
sample_mult_list = [1.2,1.5,3,3,1,.3]
fanpoint_calculator(o2_df,sample_mult_list)
o2_df.iloc[0]
o2_df[['NAME','FANPOINTS']].sort_values(by='FANPOINTS', ascending = False)

/tmp/ipykernel_84/3495819500.py:22: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  o2_df['FANPOINTS'] = fp


NAME  FANPOINTS
270              Nikola Jokic      38.42
154               Joel Embiid      35.09
134               Luka Doncic      34.10
118             Anthony Davis      33.93
11      Giannis Antetokounmpo      33.19
467          Domantas Sabonis      32.92
206              James Harden      32.27
177   Shai Gilgeous-Alexander      30.60
501              Jayson Tatum      28.97
85               Jimmy Butler      28.83
201         Tyrese Haliburton      28.71
146              Kevin Durant      28.30
259              LeBron James      27.81
574                Trae Young      27.75
318            Damian Lillard      27.74
372                 Ja Morant      27.73
479             Pascal Siakam      26.69
427        Kristaps Porzingis      26.63
412                Chris Paul      26.30
525             Fred VanVleet      26.17
22                LaMelo Ball      25.98
132         Spencer Dinwiddie      25.91
252         Jaren Jackson Jr.      25.70
103               Nic Claxton      25.56
249              Kyrie Irving      25.29
564           Zion Williamson      25.14
438             Julius Randle      25.07
179               Rudy Gobert      24.67
2                 Bam Adebayo      24.58
596        Shaquille Harrison      24.53
314             Kawhi Leonard      24.36
606              RaiQuan Gray      24.33
190            Draymond Green      24.30
116             Stephen Curry      24.20
366               Evan Mobley      24.16
529            Nikola Vucevic      24.02
474            Alperen Sengun      24.00
174               Paul George      23.96
517        Karl-Anthony Towns      23.82
548         Russell Westbrook      23.73
125             DeMar DeRozan      23.71
520              Myles Turner      23.68
231              Jrue Holiday      23.30
1                Steven Adams      23.24
420              Jakob Poeltl      23.09
248              Kyrie Irving      23.06
150           Anthony Edwards      22.99
417             Mason Plumlee      22.87
549         Russell Westbrook      22.87
607             Jacob Gilyard      22.79
30             Scottie Barnes      22.77
147              Kevin Durant      22.74
176               Josh Giddey      22.65
378           Dejounte Murray      22.56
8               Jarrett Allen      22.52
247            Brandon Ingram      22.48
57               Devin Booker      22.45
166              De'Aaron Fox      22.40
216                 Josh Hart      22.33
595               Skylar Mays      22.31
335           Lauri Markkanen      22.19
397              Jusuf Nurkic      22.10
423          Kevin Porter Jr.      22.07
289            Walker Kessler      22.00
453         Mitchell Robinson      21.97
25             Paolo Banchero      21.87
170            Darius Garland      21.85
114           Cade Cunningham      21.73
10              Kyle Anderson      21.64
323               Brook Lopez      21.59
419              Jakob Poeltl      21.56
215                 Josh Hart      21.56
90               Clint Capela      21.56
365          Donovan Mitchell      21.52
77              Jalen Brunson      21.44
481               Ben Simmons      21.11
145                 Kris Dunn      21.08
18              Deandre Ayton      20.95
73               Jaylen Brown      20.88
181              Aaron Gordon      20.60
277                 Tre Jones      20.49
39               Bradley Beal      20.39
66              Mikal Bridges      20.23
167            Markelle Fultz      20.17
522         Jonas Valanciunas      20.16
379              Jamal Murray      20.11
576               Ivica Zubac      20.02
322              Kevon Looney      20.00
107               Mike Conley      19.93
347               CJ McCollum      19.83
92         Wendell Carter Jr.      19.82
65              Mikal Bridges      19.78
485              Marcus Smart      19.71
26               Desmond Bane      19.69
406              Kelly Olynyk      19.63
463          D'Angelo Russell      19.41
570            Christian Wood      19.40
461    

In [496]:
import pandas as pd

#raw data as a csv file
filepath = 'NBASTATS2023.csv' 

#display all rows/columns
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)
rankings = pd.read_csv('data/FantasyPros_2023_Overall_NBA_ADP_Rankings.csv')

In [497]:
print(rankings.shape)
rankings.iloc[1]
rankings[['Player','AVG']].sort_values(by='AVG', ascending = True)

(258, 8)


Player    AVG
0                Nikola Jokic    1.0
1                 Luka Doncic    2.7
2       Giannis Antetokounmpo    3.3
3                 Joel Embiid    4.0
4                Jayson Tatum    4.7
5     Shai Gilgeous-Alexander    5.3
6               Stephen Curry    8.0
7           Tyrese Haliburton    8.3
8              Damian Lillard    9.3
9                Kevin Durant   10.7
10           Domantas Sabonis   12.0
11              Anthony Davis   14.0
12            Anthony Edwards   14.3
13                 Trae Young   14.3
14                LaMelo Ball   15.3
15               Devin Booker   15.7
16               LeBron James   16.3
17           Donovan Mitchell   19.0
18          Victor Wembanyama   20.0
19               Kyrie Irving   21.3
20              Mikal Bridges   21.7
21              Pascal Siakam   24.0
22               Jimmy Butler   25.0
23         Karl-Anthony Towns   26.7
24               De'Aaron Fox   29.0
25                Bam Adebayo   29.3
26            Lauri Markkanen   30.7
27            Dejounte Murray   32.3
28              Jalen Brunson   33.0
29          Jaren Jackson Jr.   33.3
30             Nikola Vucevic   33.3
31               Jaylen Brown   33.7
32              DeMar DeRozan   35.7
33              Kawhi Leonard   36.0
34              Julius Randle   37.0
35             Darius Garland   37.3
36            Zion Williamson   37.7
37               James Harden   37.7
38         Kristaps Porzingis   38.3
39               Jordan Poole   38.7
40             Brandon Ingram   40.3
41               Desmond Bane   41.0
42            Cade Cunningham   41.3
43               Jamal Murray   41.3
44              Fred VanVleet   42.0
45                Paul George   42.0
46             Paolo Banchero   42.7
47                Zach LaVine   45.0
48                Evan Mobley   47.3
49             Scottie Barnes   50.0
50               Jrue Holiday   51.7
51              Chet Holmgren   52.7
52                Josh Giddey   53.0
53              Deandre Ayton   54.0
54               Tyrese Maxey   56.7
55                  Ja Morant   58.3
56               Myles Turner   59.7
57               Bradley Beal   61.0
58             Alperen Sengun   61.0
59             Walker Kessler   64.0
60                 Kyle Kuzma   64.3
61              Jarrett Allen   66.0
62               Franz Wagner   66.0
63                Rudy Gobert   66.3
64                CJ McCollum   67.3
65                 Tyus Jones   68.7
66                Tyler Herro   70.0
67            Khris Middleton   70.3
68                 OG Anunoby   71.0
69            Scoot Henderson   72.0
70             Jalen Williams   72.3
71              Klay Thompson   72.7
72            Anfernee Simons   74.3
73                Nic Claxton   74.7
74                Jalen Green   76.3
75              Austin Reaves   77.3
76      Russell Westbrook III   78.3
77               Jerami Grant   78.3
78              Devin Vassell   79.7
79               Terry Rozier   81.3
80          Spencer Dinwiddie   82.3
81          Jonas Valanciunas   82.3
82             Andrew Wiggins   82.3
83               Jakob Poeltl   83.0
84              Tobias Harris   84.3
85                Brook Lopez   86.7
86                 Chris Paul   86.7
87               Clint Capela   89.3
88            Cameron Johnson   90.3
89           Jabari Smith Jr.   90.3
90           D'Angelo Russell   93.3
91         Michael Porter Jr.   94.0
92               Marcus Smart   94.7
93                Buddy Hield   95.0
94               John Collins   95.7
95               Jusuf Nurkic   96.3
96                Ben Simmons   97.0
97              Keegan Murray   97.0
98             Keldon Johnson   97.0
99         Wendell Carter Jr.  100.7
100            Draymond Green  101.7
101            Markelle Fultz  103.0
102                RJ Barrett  105.7
103           Jordan Clarkson  105.7
104              Aaron Gordon  106.3
105             Derrick White  107.0
106          Bobby Portis Jr.  107.3
107             Mark Willia

This is a table that contains the average pick each player is taken in fantasy drafts across different websites.

In fantasy basketball, higher average draft pick means that the player has higher value when you are picking(opponent may select). Lets combine our rankings and stats dataframes and create a new column for weighted fantasy value(Fantasy points per game /(dp_multiplier ** AVG draft pick). dp_multiplier slightly >1 because a earlier avg draft pick with same points is slightly higher priority.

We will do an inner join on the name of the player to combine our data from our two tables, and then add our final column, WFPG(Weighted Fan Points/Game)

In [ ]:
#prepares a rankings dataframe and a dataframe passed through fanpoint_calculator
#for the get WFPG method, and returns a merged dataframe
#df1: rankings dataframe downloaded from fantasypros formatted csv
#df2: dataframe passed through fanpoint calculator
#result: merged dataframe of df1 and df2 inner joined on player name
def prepare(df1, df2):
    df1 = df1.rename(columns = {'Player':'NAME'})
    df2 = df2.drop_duplicates('NAME')
    df1 = df1.drop_duplicates('NAME')
    result = df1.merge(df2, how = 'inner', on = ['NAME'])
    return result

In [570]:
#calculates weighted fan points/game for each player and returns a dataframe with WFPG
#column added
#df1: rankings ddataframe from fantasypros formatted csv
#df2: dataframe passed through fanpoint calculator
#dp_multiplier: exponential multiplier for draft pick
def get_WFPG(df1, df2, dp_multiplier):
    result = prepare(df1,df2)
    weightedFP = []
    for index,row in result.iterrows():
        wfpg = ((row['GP']/82)*(row['FANPOINTS']))/(dp_multiplier **row['AVG'])
        weightedFP.append(wfpg)
    
    result['WFPG'] = weightedFP
    
    return result


In [578]:

result = get_WFPG(rankings, weighted, 1.002)
print(result.shape)
result[['NAME','WFPG','GP']].sort_values(by='WFPG', ascending = False)

(217, 36)
(217, 37)


NAME       WFPG  GP
0                Nikola Jokic  32.264495  69
10           Domantas Sabonis  30.964239  79
3                 Joel Embiid  28.018351  66
1                 Luka Doncic  27.298678  66
4                Jayson Tatum  25.899303  74
2       Giannis Antetokounmpo  25.332058  63
5     Shai Gilgeous-Alexander  25.108315  68
13                 Trae Young  24.008420  73
11              Anthony Davis  22.532529  56
29             Nikola Vucevic  22.473871  82
20              Pascal Siakam  22.027629  71
33              Julius Randle  21.863793  77
12            Anthony Edwards  21.525031  79
21               Jimmy Butler  21.405123  64
24                Bam Adebayo  21.203378  75
47                Evan Mobley  21.177122  79
36               James Harden  21.168984  58
69                Nic Claxton  20.405244  76
43              Fred VanVleet  20.248584  69
31              DeMar DeRozan  19.923781  74
37         Kristaps Porzingis  19.554060  65
56             Alperen Sengun  19.432454  75
48             Scottie Barnes  19.348791  77
7           Tyrese Haliburton  19.284362  56
8              Damian Lillard  19.259756  58
26            Dejounte Murray  19.086643  74
50                Josh Giddey  18.883335  76
23               De'Aaron Fox  18.818851  73
28          Jaren Jackson Jr.  18.474160  63
61                Rudy Gobert  18.446900  70
53                  Ja Morant  18.360212  61
16               LeBron James  18.055352  55
95             Draymond Green  17.655011  73
45             Paolo Banchero  17.632574  72
57             Walker Kessler  17.470542  74
80                Brook Lopez  17.270382  78
17           Donovan Mitchell  17.181086  68
49               Jrue Holiday  17.169422  67
34             Darius Garland  17.065565  69
25            Lauri Markkanen  16.797643  66
27              Jalen Brunson  16.645047  68
76          Jonas Valanciunas  16.477461  79
46                Zach LaVine  16.410329  77
59              Jarrett Allen  16.367933  68
6               Stephen Curry  16.264764  56
30               Jaylen Brown  15.949579  67
81                 Chris Paul  15.913381  59
54               Myles Turner  15.891195  62
62                CJ McCollum  15.855227  75
51              Deandre Ayton  15.366962  67
121              Kevon Looney  15.293081  82
38               Jordan Poole  15.087135  82
44                Paul George  15.045848  56
42               Jamal Murray  14.678263  65
66             Jalen Williams  14.581452  75
109            Onyeka Okongwu  14.537792  80
60               Franz Wagner  14.450844  80
32              Kawhi Leonard  14.375693  52
119               Ivica Zubac  14.367971  76
100             Derrick White  14.317346  82
82               Clint Capela  14.297521  65
15               Devin Booker  14.062260  53
99               Aaron Gordon  13.814128  68
84           Jabari Smith Jr.  13.730663  79
102                 Tre Jones  13.674633  68
63                 Tyus Jones  13.573647  80
70                Jalen Green  13.345217  76
9                Kevin Durant  13.175059  39
88                Buddy Hield  13.161232  80
79              Tobias Harris  13.047249  74
141             Kyle Anderson  13.035350  69
58                 Kyle Kuzma  12.979655  64
19              Mikal Bridges  12.935133  56
40               Desmond Bane  12.831673  58
74               Terry Rozier  12.650583  63
64                Tyler Herro  12.574586  67
124         Immanuel Quickley  12.522899  81
107                Jaden Ivey  12.464931  74
111         Mitchell Robinson  12.388142  59
115            Christian Wood  12.340625  67
89               John Collins  12.229245  71
156          Kevin Porter Jr.  12.149707  59
87               Marcus Smart  12.134709  61
72               Jerami Grant  12.043329  63
96             Markelle Fultz  12.013456  60
188             Killian Hayes  11.997716  76
105               Jalen Duren  11.727102  67
152            T.J. McConnell  11.661997  75
125         De'Antho

Our rankings data contains 258 players, and not all of them match the data in our previous dataframe so in this case where our data is mismatched size, we will compute weighted fan point values only for those players that are contained in both datasets(217).  

The index for each player shows their original placement based on the rankings data(0 based indexing), we can see how rankings have shifted around based on our league's point calculations

We have now merged our dataframes, and created an effective ranking system for my fantasy draft!
